src/core/server/logging/appenders/rolling_file/rolling_file_appender.ts

```diff
@@ -0,0 +1,174 @@
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { schema } from '@kbn/config-schema';
+import { LogRecord, Layout, DisposableAppender } from '@kbn/logging';
+import { Layouts, LayoutConfigType } from '../../layouts/layouts';
+import { BufferAppender } from '../buffer/buffer_appender';
+import {
+  TriggeringPolicyConfig,
+  createTriggeringPolicy,
+  triggeringPolicyConfigSchema,
+  TriggeringPolicy,
+} from './policies';
+import {
+  RollingStrategy,
+  createRollingStrategy,
+  RollingStrategyConfig,
+  rollingStrategyConfigSchema,
+} from './strategies';
+import { RollingFileManager } from './rolling_file_manager';
+import { RollingFileContext } from './rolling_file_context';
+
+export interface RollingFileAppenderConfig {
+  kind: 'rolling-file';
+  /**
+   * The layout to use when writing log entries
+   */
+  layout: LayoutConfigType;
+  /**
+   * The absolute path of the file to write to.
+   */
+  path: string;
+  /**
+   * The {@link TriggeringPolicy | policy} to use to determine if a rollover should occur.
+   */
+  policy: TriggeringPolicyConfig;
+  /**
+   * The {@link RollingStrategy | rollout strategy} to use for rolling.
+   */
+  strategy: RollingStrategyConfig;
+}
+
+/**
+ * Appender that formats all the `LogRecord` instances it receives and writes them to the specified file.
+ * @internal
+ */
+export class RollingFileAppender implements DisposableAppender {
+  public static configSchema = schema.object({
+    kind: schema.literal('rolling-file'),
+    layout: Layouts.configSchema,
+    path: schema.string(),
+    policy: triggeringPolicyConfigSchema,
+    strategy: rollingStrategyConfigSchema,
+  });
+
+  private isRolling = false;
+  private disposed = false;
+  private rollingPromise?: Promise<void>;
+
+  private readonly layout: Layout;
+  private readonly context: RollingFileContext;
+  private readonly fileManager: RollingFileManager;
+  private readonly policy: TriggeringPolicy;
+  private readonly strategy: RollingStrategy;
+  private readonly buffer: BufferAppender;
+
+  constructor(config: RollingFileAppenderConfig) {
+    this.context = new RollingFileContext(config.path);
+    this.context.refreshFileInfo();
+    this.fileManager = new RollingFileManager(this.context);
+    this.layout = Layouts.create(config.layout);
+    this.policy = createTriggeringPolicy(config.policy, this.context);
+    this.strategy = createRollingStrategy(config.strategy, this.context);
+    this.buffer = new BufferAppender();
+  }
+
+  /**
+   * Formats specified `record` and writes it to the specified file. If the record
+   * would trigger a rollover, it will be performed before the effective write operation.
+   */
+  public append(record: LogRecord) {
```

Comment:
Not sure to understand,

Internally, once disposed, we will never call `append`. Even if some messages where queued into the buffer, we'll directly write them to file (`_writeToFile`) instead of calling `append` (to avoid the risk of triggering another rolling when the appender is disposed).

Externally, nothing is currently forbidding to call `append` once the appender is disposed, but that's something already present in our other appenders (you can call `append` on `FileAppender` once disposed), it's the logging system responsibility to avoid such scenarios.

Reply:
>  but that's something already present in our other appenders 

it wasn't the problem because `dispose` used to be synchronous. Ok, it shouldn't block `RollingFile` appender merge